## learning-AI101 : topic2_2

topic2_1에서는 **keras를 이용하여 모델**을 생성하였다.  
만든 모델을 **model.fit을 하기 위해선 X_train, y_train, epochs값이 필요**하다.  
**CSV 데이터를 전처리하기 위해선 pandas가 필요하다.**

In [1]:
import tensorflow as tf
import pandas as pd

2024-01-17 13:39:43.280463: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [5]:
# 학습 모델 만든 것

model = tf.keras.models.Sequential([
    tf.keras.layers.Dense(64, activation="relu"), 
    tf.keras.layers.Dense(128, activation="relu"),
    tf.keras.layers.Dense(1, activation="sigmoid")])

model.compile(optimizer="adam", loss="binary_crossentropy", metrics=['accuracy'])

# model.fit(x데이터, y데이터, epochs=에폭값)

--------

model.fit에 들어갈 데이터는 아래와 같다.
- X_train : 정답 예측에 필요한 input data
- y_train : X_train의 정답 (label)
- epochs : 학습 반복 수

X_train는 보통 **2차원 리스트 데이터 혹은 metrics** 형태이다.

In [9]:
dataframe = pd.read_csv("./data/gpascore.csv")
dataframe

admit    gre   gpa  rank
0        0  380.0  3.21     3
1        1  660.0  3.67     3
2        1  800.0  4.00     1
3        1  640.0  3.19     4
4        0  520.0  2.33     4
..     ...    ...   ...   ...
421      1  750.0  3.44     4
422      1  710.0  3.66     4
423      0  450.0  3.25     4
424      1  760.0  3.76     2
425      1  710.0  3.82     3

[426 rows x 4 columns]

X_train은 gre, gpa, rank이며 y_train은 admit일 것이다.  
X_train의 형식은 **[[380.0, 3.21, 3], [660.0, 3.67, 3] ...]이다.**  
y_train의 형식은 **[0, 1, 1, 1, 0...]이다.**

In [13]:
# pandas를 이용하여 NULL값을 빼주는 전처리를 해보자
print(dataframe.isnull().sum()) # NULL값 갯수 세주기
# gre라는 column에 NULL값이 하나 있다

admit    0
gre      1
gpa      0
rank     0
dtype: int64


In [29]:
# NULL값이 있는 row를 없애는 함수 (전처리 완료)
dataframe = dataframe.dropna()
dataframe.shape
# 원래 426개인 row가 425개가 되었다
# # data.fillna(~~) : ~~를 NULL값이 있는 모든 곳에 채워줌

(425, 4)

In [30]:
# pandas 유용한 기능
# 원하는 column만 출력
dataframe['admit']

0      0
1      1
2      1
3      1
4      0
      ..
421    1
422    1
423    0
424    1
425    1
Name: admit, Length: 425, dtype: int64

In [31]:
# 특정 column.min() or column.max() : 그 열의 최솟값 or 최댓값 리턴
# 특정 column.mean : 그 열의 평균 리턴
# 특정 column.count() : 그 열의 원소 갯수
print (dataframe['gpa'].min())
print (dataframe['gpa'].max())
print (dataframe['gpa'].mean())
print (dataframe['gpa'].count())

2.12
4.0
3.385929411764706
425


In [50]:
# X_train, y_train 만들기
y_train = dataframe['admit'].values # .values : column 값들을 1차원 리스트로 담아줌

X_train = []
for i, rows in dataframe.iterrows() : 
    # iterrow() : df의 한 행씩 출력
    # i : index, rows : 한 행
    X_train.append ([rows['gre'], rows['gpa'], rows['rank']])

print (X_train)

[[380.0, 3.21, 3.0], [660.0, 3.67, 3.0], [800.0, 4.0, 1.0], [640.0, 3.19, 4.0], [520.0, 2.33, 4.0], [760.0, 3.0, 2.0], [560.0, 2.98, 1.0], [400.0, 3.08, 2.0], [540.0, 3.39, 3.0], [700.0, 3.92, 2.0], [750.0, 3.6, 4.0], [440.0, 3.22, 1.0], [760.0, 4.0, 1.0], [600.0, 3.08, 2.0], [700.0, 4.0, 1.0], [480.0, 3.44, 3.0], [490.0, 3.87, 4.0], [360.0, 2.56, 3.0], [800.0, 3.75, 2.0], [500.0, 3.17, 3.0], [660.0, 3.63, 2.0], [600.0, 2.82, 4.0], [680.0, 3.19, 4.0], [760.0, 3.35, 2.0], [800.0, 3.66, 1.0], [620.0, 3.61, 1.0], [520.0, 3.74, 4.0], [780.0, 3.22, 2.0], [520.0, 3.29, 1.0], [540.0, 3.78, 4.0], [760.0, 3.35, 3.0], [600.0, 3.4, 3.0], [800.0, 4.0, 3.0], [360.0, 3.14, 1.0], [400.0, 3.05, 2.0], [580.0, 3.25, 1.0], [520.0, 2.9, 3.0], [500.0, 3.13, 2.0], [520.0, 2.68, 3.0], [560.0, 2.42, 2.0], [580.0, 3.32, 2.0], [600.0, 3.15, 2.0], [500.0, 3.31, 3.0], [700.0, 2.94, 2.0], [460.0, 3.45, 3.0], [580.0, 3.46, 2.0], [500.0, 2.97, 4.0], [440.0, 2.48, 4.0], [400.0, 3.35, 3.0], [640.0, 3.86, 3.0], [440.0,